algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py

```diff
@@ -0,0 +1,730 @@
+                                 dim)
+    self.ra_var = self.variable('batch_stats',
+                                'var',
+                                lambda s: jnp.ones(s, dtype),
+                                dim)
+
+    self.gamma = self.param('scale', nn.initializers.zeros, dim, dtype)
+    self.beta = self.param('bias', nn.initializers.zeros, dim, dtype)
+
+  def _get_default_paddings(self, inputs):
+    """Gets the default paddings for an input."""
+    in_shape = list(inputs.shape)
+    in_shape[-1] = 1
+
+    return jnp.zeros(in_shape, dtype=inputs.dtype)
+
+  @nn.compact
+  def __call__(self, inputs, input_paddings=None, train=False):
+    rank = inputs.ndim
+    reduce_over_dims = list(range(0, rank - 1))
+
+    if input_paddings is None:
+      padding = self._get_default_paddings(inputs)
+    else:
+      padding = jnp.expand_dims(input_paddings, -1)
+
+    momentum = self.batch_norm_momentum
+    epsilon = self.batch_norm_epsilon
+
+    if train:
+      mask = 1.0 - padding
+      sum_v = jnp.sum(inputs * mask, axis=reduce_over_dims, keepdims=True)
+      count_v = jnp.sum(
+          jnp.ones_like(inputs) * mask, axis=reduce_over_dims, keepdims=True)
+
+      count_v = jnp.maximum(count_v, 1.0)
+      mean = sum_v / count_v
+
+      sum_vv = jnp.sum(
+          (inputs - mean) * (inputs - mean) * mask,
+          axis=reduce_over_dims,
+          keepdims=True)
+
+      var = sum_vv / count_v
+
+      self.ra_mean.value = momentum * self.ra_mean.value + (1 - momentum) * mean
+      self.ra_var.value = momentum * self.ra_var.value + (1 - momentum) * var
+    else:
+      mean = self.ra_mean.value
+      var = self.ra_var.value
+
+    inv = (1 + self.gamma) / jnp.sqrt(var + epsilon)
+
+    bn_output = (inputs - mean) * inv + self.beta
+    bn_output *= 1.0 - padding
+
+    return bn_output
+
+
+@jax.vmap
+def flip_sequences(inputs: Array, lengths: Array) -> Array:
+  """Flips a sequence of inputs along the time dimension.
+
+  This function can be used to prepare inputs for the reverse direction of a
+  bidirectional LSTM. It solves the issue that, when naively flipping multiple
+  padded sequences stored in a matrix, the first elements would be padding
+  values for those sequences that were padded. This function keeps the padding
+  at the end, while flipping the rest of the elements.
+
+  Example:
+  ```python
+  inputs = [[1, 0, 0],
+            [2, 3, 0]
+            [4, 5, 6]]
+  lengths = [1, 2, 3]
+  flip_sequences(inputs, lengths) = [[1, 0, 0],
+                                     [3, 2, 0],
+                                     [6, 5, 4]]
+  ```
+
+  Args:
+    inputs: An array of input IDs <int>[batch_size, seq_length].
+    lengths: The length of each sequence <int>[batch_size].
+
+  Returns:
+    An ndarray with the flipped inputs.
+  """
+  # Compute the indices to put the inputs in flipped order as per above example.
+  max_length = inputs.shape[0]
```

Comment:
instead of taking `lengths` as an input, I think we can recompute it as `lengths = jnp.argmin(inputs, axis=1)`

Reply:
made the change to pass input_paddings as deep as possible